python/ray/data/impl/pandas_block.py

```diff
@@ -0,0 +1,189 @@
+        view = self._table[start:end]
+        if copy:
+            view = view.copy(deep=True)
+        return view
+
+    def random_shuffle(self, random_seed: Optional[int]) -> List[T]:
+        return self._table.sample(frac=1, random_state=random_seed)
+
+    def schema(self) -> Any:
+        dtypes = self._table.dtypes
+        schema = PandasBlockSchema(
+            names=dtypes.index.tolist(), types=dtypes.values.tolist())
+        # Column names with non-str types of a pandas DataFrame is not
+        # supported by Ray Dataset.
+        if any(not isinstance(name, str) for name in schema.names):
+            raise ValueError(
+                "A Pandas DataFrame with column names of non-str types"
+                " is not supported by Ray Dataset. Column names of this"
+                f" DataFrame: {schema.names!r}.")
+        return schema
+
+    def to_pandas(self) -> "pandas.DataFrame":
+        return self._table
+
+    def to_numpy(self, column: str = None) -> np.ndarray:
+        if not column:
+            raise ValueError(
+                "`column` must be specified when calling .to_numpy() "
+                "on Pandas blocks.")
+        if column not in self._table.columns:
+            raise ValueError(
+                "Cannot find column {}, available columns: {}".format(
+                    column, self._table.columns.tolist()))
+        return self._table[column].to_numpy()
+
+    def to_arrow(self) -> "pyarrow.Table":
+        import pyarrow
+        return pyarrow.table(self._table)
+
+    def num_rows(self) -> int:
+        return self._table.shape[0]
+
+    def size_bytes(self) -> int:
+        # TODO (kfstorm): Should we count index?
+        return self._table.memory_usage(index=True, deep=True).sum()
+
+    def _zip(self, acc: BlockAccessor) -> "Block[T]":
+        r = self.to_pandas().copy(deep=False)
+        s = acc.to_pandas()
+        for col_name in s.columns:
+            col = s[col_name]
+            # Ensure the column names are unique after zip.
+            if col_name in r.column_names:
+                i = 1
+                new_name = col_name
+                while new_name in r.column_names:
+                    new_name = "{}_{}".format(col_name, i)
+                    i += 1
+                col_name = new_name
+            r[col_name] = col
+        return r
+
+    @staticmethod
+    def builder() -> PandasBlockBuilder[T]:
+        return PandasBlockBuilder()
+
+    @staticmethod
+    def _empty_table() -> "pandas.DataFrame":
+        return PandasBlockBuilder._empty_table()
+
+    def _sample(self, n_samples: int, key: SortKeyT) -> "pandas.DataFrame":
+        return self._table[[k[0] for k in key]].sample(
+            n_samples, ignore_index=True)
+
+    def sort_and_partition(self, boundaries: List[T], key: SortKeyT,
+                           descending: bool) -> List["Block[T]"]:
+        # TODO (kfstorm): A workaround to pass tests. Not efficient.
```

Comment:
We should be sure to open some P2 issues for adding Pandas-native implementations of these!

Reply:
Thanks for the heads up! Issue created. #21296